gtk: Support GTK 3.18

jhenstridge opened the pull request:
This is an initial pull request towards supporting GTK 3.18 as found in Ubuntu 16.04 (and snaps based on that distro release running on newer systems).

The diff is a bit messy because it moves some directories around.  Namely:

1. move `gtk/src/light/gtk-3.0` to `gtk/src/light/gtk-3.20`
2. copy GTK 3.18's Adwaita to `gtk/src/light/gtk-3.0`
3. try to adapt the 3.18 theme to look like the 3.20 theme.

It is quite easy to rebase this on changes to master, but could make updating other pull requests difficult.  So I'm not sure when would be best to merge it.

Both versions of the theme are installed side by side so new GTK will use the main version, and old GTK will use the fallback.

There is still work to do, such as porting the image assets over, but it is already an improvement.  Before:
![image](https://user-images.githubusercontent.com/12724165/47553695-498b5a00-d93a-11e8-873b-78da535de4c5.png)

After:
![image](https://user-images.githubusercontent.com/12724165/47553750-6d4ea000-d93a-11e8-91a1-f62385e80141.png)

One thing left to do that would have the most dramatic effect would be to replace the image assets.  I haven't done this yet, since (a) the SVG asset sheet for new and old Adwaita is quite different (check and radio buttons are composed differently), and (b) the Yaru assets don't seem to bear any relation to the asset sheet in master.

Things also look a bit different due to GTK 3.18 missing support for a few CSS properties.  Most notably `min-width`/`min-height`.  In particular, this affects the size of the minimise/maximise/close buttons in the header.

clobrano commented:
Hi @jhenstridge, thanks for this work

> It is quite easy to rebase this on changes to master, but could make updating other pull requests difficult. So I'm not sure when would be best to merge it.

after the queue of current PRs would be a good time?

> (b) the Yaru assets don't seem to bear any relation to the asset sheet in master.

could you clarify this? Is there anything we can do to help?

jhenstridge commented:
> after the queue of current PRs would be a good time?
I only brought this up because git might get confused merging the other PRs after landing this one.  As I've moved a bunch of files and added new files with the same names in the old locations, git may attempt to merge with the wrong set of files.

On the other hand, rebasing this branch after other PRs have been merged is relatively painless.  I wouldn't like to hold off on this indefinitely though, since it will be easier to merge incremental improvements after the bulk goes in.

> could you clarify this? Is there anything we can do to help?
I think I got a bit confused since there are e.g. `assets/checkbox-checked-selectionmode.png` that doesn't come from the asset sheet.  I see now that it is constructed via CSS with the only asset being the checkmark.

For Adwaita 3.18, all the checkbox and radio button states are handled as PNGs in the asset sheet.  I'll see if I can get them ported over without having to produce a new asset sheet.

jhenstridge commented:
Okay, I've updated the asset sheet with sliders and text select markers from the GTK 3.20 theme, and check and radio images from the GTK 2 theme, with colours for the extra states taken from the 3.20 generated CSS.

I originally tried reusing the CSS based method of constructing checkboxes and radio buttons, but it seems 3.18 doesn't do the calls necessary to render borders or backgrounds on the indicator.

The main obvious differences now are:
1. highlight colour for sliders and progress bars is orange rather than blue
2. missing the orange underline for selected GtkStack page
3. padding is off on notebook header
4. window control buttons are too small

I should be able to sort out the first two tomorrow.  I haven't had much success trying to solve the last two.

jhenstridge commented:
And here's the current progress:

![image](https://user-images.githubusercontent.com/12724165/47699850-3389dc00-dc4f-11e8-93e3-2e7204ee9f11.png)

I got the title buttons to about the right size via padding.  Scales, progress bars, level bars, and spinners are now blue.  Check boxes, radio buttons, and scales use Yaru assets from other versions of the theme.  On that page, the most obvious difference is switches now.

Feichtmeier commented:
That looks really good, thank you!
What are those apps that use gtk 3.18? Just curious :)

jhenstridge commented on gtk/src/light/gtk-3.0/meson.build:
```diff
@@ -16,117 +16,368 @@ gtk3_scss_dependencies = [
   'assets/checkbox-backdrop-selectionmode.png',
+  'assets/checkbox-checked@2.png',
+  'assets/checkbox-checked-active@2.png',
+#  'assets/checkbox-checked-active-dark@2.png',
```
These are assets that Adwaita 3.18 used I pulled out when adapting for Yaru.  There's basically two cases:

1. cases where we're using the light assets in dark mode.
2. Adwaita had slightly different versions of check boxes/radio buttons for use in selected list rows, where I just used the standard versions.

I don't know what the status will be for all these assets going forward (or whether we'd even want to bother improving the 3.18 theme going forward).  What do you think?

didrocks replied:
TBH, you own it I think, so just minimize future 3-way diff check (I think remove them)

jhenstridge replied:
I've removed the unused images, and trimmed `assets.txt` to just those we actually want to export now.  Commented lines in the `meson.build` are gone too.

didrocks commented:
> What are those apps that use gtk 3.18? Just curious :)
@Feichtmeier: every gtk applications using gtk =< 3.18. You can think of people using the Yaru theme on 16.04 for instance! :)

jhenstridge commented:
> What are those apps that use gtk 3.18? Just curious :)

The primary use case is snap applications built against Ubuntu 16.04 libraries.  We can provide the Yaru theme to them via the `gtk-common-themes` snap, but it needs to be compatible with the libgtk they're using.

jhenstridge commented on gtk/src/dark/gtk-3.0/meson.build:
```diff
@@ -5,18 +5,18 @@ gtk3_asset_dir = join_paths(gtk3_dir, 'assets')
-  '../../light/gtk-3.0/_common.scss',
-  '../../light/gtk-3.0/_apps.scss',
-  '../../light/gtk-3.0/_drawing.scss',
+  '../../light/gtk-3.20/_colors-public.scss',
```
I've updated the dark theme build to produce `gtk-3.0` and `gtk-3.20` dirs.  Also, rather than rebuilding the CSS, I changed it to just symlink to the light theme's `gtk-dark.css`.

jhenstridge commented:
Okay, I think I've resolved all the issues @didrocks brought up.  I fixed up the switches, and got rid of the unwanted text/icon shadows on the headerbar buttons.  That gives us a somewhat passable theme now:

![image](https://user-images.githubusercontent.com/12724165/47786939-5a780900-dd48-11e8-9761-454cdc041fc8.png)

didrocks commented:
This looks good to me! Thanks for the changes. I trust that you tested the snap and there is no surprise with the symlinks as side effect.

Planning on merging on Monday! :+1: for me.

jhenstridge commented:
> This looks good to me! Thanks for the changes. I trust that you tested the snap and there is no surprise with the symlinks as side effect.

Well, the `snapcraft.yaml` override-build` script deletes the dark theme, so it can't make any difference there.  The links seem to work fine in local testing of the theme though.

didrocks commented:
Thanks a lot! Merging now :)

clobrano commented:
Hi @jhenstridge, 

sorry to bring to light this again. I am about to finish the details to merge gtk2 dark version in master, my understanding is that, except the build scripts, the changes in this PR do not affect gtk2, is that right?
